### PR TITLE
sriov: Add failover cases

### PR DESCRIPTION
**Pull request by Yingshun**
Add 4 failover cases:
    VIRT-293923 - [failover] Hotunplug and then hotplug the hostdev
    interface with failover setting
    VIRT-293926 - [failover] lifecycle test for vm with failover setting
    VIRT-293911 - [failover] Cancel vm migration with failover settings
    VIRT-293874: [failover] Migrate vm with failover setting

BTW, add vmxml backup to SRIOVTest().

Signed-off-by: Yingshun Cui <yicui@redhat.com>
**depends on:** https://github.com/avocado-framework/avocado-vt/pull/3448
**Test results:** 2 cases are failed due to a known issue.
```
 (1/4) type_specific.io-github-autotest-libvirt.sriov.failover.migration.hostdev_interface.with_precopy: PASS (372.70 s)
 (2/4) type_specific.io-github-autotest-libvirt.sriov.failover.migration.hostdev_interface.with_postcopy: PASS (350.82 s)
 (3/4) type_specific.io-github-autotest-libvirt.sriov.failover.migration.hostdev_device.with_precopy: PASS (359.02 s)
 (4/4) type_specific.io-github-autotest-libvirt.sriov.failover.migration.hostdev_device.with_postcopy: PASS (325.99 s)
 (1/2) type_specific.io-github-autotest-libvirt.sriov.failover.abort_migration.hostdev_interface.with_precopy: PASS (253.63 s)
 (2/2) type_specific.io-github-autotest-libvirt.sriov.failover.abort_migration.hostdev_device.with_precopy: PASS (230.60 s)
 (1/2) type_specific.io-github-autotest-libvirt.sriov.failover.at_dt_hostdev.hostdev_interface: PASS (102.42 s)
 (2/2) type_specific.io-github-autotest-libvirt.sriov.failover.at_dt_hostdev.hostdev_device: PASS (101.45 s)
 (1/2) type_specific.io-github-autotest-libvirt.sriov.failover.lifecycle.hostdev_interface: ERROR: VM 'avocado-vt-vm1' failed to start: error: Failed to start domain 'avocado-vt-vm1'\nerror: Cannot get interface MTU on 'br0': No such device (7.42 s)
 (2/2) type_specific.io-github-autotest-libvirt.sriov.failover.lifecycle.hostdev_device: ERROR: VM 'avocado-vt-vm1' failed to start: error: Failed to start domain 'avocado-vt-vm1'\nerror: Cannot get interface MTU on 'br0': No such device (7.57 s)

```

### CODE REVIEW

**Comment by Yingshun** [comment]
```
 (01/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_interface.vf_address.managed_yes.with_iommu: PASS (156.86 s)
 (02/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_interface.vf_address.managed_yes.without_iommu: PASS (40.95 s)
 (03/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_interface.vf_address.managed_no.without_iommu: PASS (40.52 s)
 (04/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_interface.vf_address.without_managed.without_iommu: PASS (40.18 s)
 (05/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_interface.vf_address.vlan.without_iommu: PASS (38.17 s)
 (06/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_interface.vf_address.failover.with_iommu: PASS (185.83 s)
 (07/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_device.vf_address.managed_yes.with_iommu: PASS (156.23 s)
 (08/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_device.vf_address.managed_yes.without_iommu: PASS (39.75 s)
 (09/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_device.vf_address.managed_no.without_iommu: PASS (40.05 s)
 (10/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_device.pf_address.managed_yes.without_iommu: PASS (41.15 s)
 (11/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.hostdev_device.pf_address.managed_no.without_iommu: PASS (43.54 s)
 (12/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.network_interface.network.pf_name.hostdev.managed_no.without_iommu: PASS (42.53 s)
 (13/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.network_interface.network.pf_name.hostdev.managed_yes.without_iommu: PASS (40.62 s)
 (14/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.network_interface.network.pf_name.hostdev.without_managed.without_iommu: PASS (43.77 s)
 (15/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.network_interface.network.vf_address.hostdev.managed_yes.without_iommu: PASS (42.53 s)
 (16/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.suspend_resume.network_interface.network.vf_address.hostdev.without_managed.without_iommu: PASS (40.91 s)
 (17/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.hostdev_interface.vf_address.managed_no: PASS (16.44 s)
 (18/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.hostdev_interface.vf_address.without_managed: PASS (16.28 s)
 (19/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.hostdev_device.vf_address.managed_no: PASS (16.39 s)
 (20/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.hostdev_device.pf_address.managed_no: PASS (16.33 s)
 (21/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.network_interface.network.pf_name.managed_no: PASS (16.79 s)
 (22/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.network_interface.network.pf_name.without_managed: PASS (16.57 s)
 (23/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.cold_plug.network_interface.network.vf_address.without_managed: PASS (16.61 s)
 (24/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.hostdev_interface.vf_address.managed_no: PASS (18.49 s)
 (25/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.hostdev_interface.vf_address.without_managed: PASS (18.87 s)
 (26/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.hostdev_device.vf_address.managed_no: PASS (18.85 s)
 (27/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.hostdev_device.pf_address.managed_no: PASS (22.37 s)
 (28/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.network_interface.network.pf_name.managed_no: PASS (20.04 s)
 (29/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.network_interface.network.pf_name.without_managed: PASS (19.16 s)
 (30/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.unmanaged.hot_plug.network_interface.network.vf_address.without_managed: PASS (19.07 s)
 (31/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.iommu.at_dt_hostdev.hostdev_interface.managed_yes: PASS (190.42 s)
 (32/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.iommu.at_dt_hostdev.hostdev_interface.failover: PASS (208.15 s)
 (33/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.iommu.at_dt_hostdev.hostdev_device.managed_yes: PASS (192.08 s)
 (34/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_interface.vf_address.managed_yes: PASS (45.27 s)
 (35/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_interface.vf_address.managed_no: PASS (45.55 s)
 (36/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_interface.vf_address.without_managed: PASS (45.97 s)
 (37/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_interface.vf_address.vlan: PASS (42.51 s)
 (38/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_device.vf_address.managed_yes: PASS (44.82 s)
 (39/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_device.vf_address.managed_no: PASS (45.01 s)
 (40/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_device.pf_address.managed_yes: PASS (45.85 s)
 (41/64) type_specific.io-github-autotest-libvirt.sriov.vm_lifecycle.start_destroy.hostdev_device.pf_address.managed_no: PASS (47.07 s)
...

```

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        1. Start vm with failover setting interfaces
+        2. Check VM's interface quantity and network connectivity
+        3. Hot-unplug the hostdev interface/device by detach-device
+        4. After hot-unplug the hostdev interface, ensure the bridge interface
+            will be active, so check on vm for 1) interface quantity; 2)network
+            connectivity
+        5. Hot plug the hostdev interface/device back by attach-device
+        6. Check on vm for 1) interface quantity; 2)network connectivity
+        """
+        # utils_sriov.set_vf_mac(sriov_test_obj.pf_name, mac_addr)
```
Comment (cliping) [inline libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py:28]: Maybe we can remove this line.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        check_points.check_vm_network_accessed(vm_session,
+                                               tcpdump_iface=br_name,
+                                               tcpdump_status_error=True)
+
+    dev_type = params.get("dev_type", "")
+    set_vf_mac = "yes" == params.get("set_vf_mac", "no")
+    vm_name = params.get("main_vm", "avocado-vt-vm1")
+    vm = env.get_vm(vm_name)
+    sriov_test_obj = sriov_base.SRIOVTest(vm, test, params)
+    # default_vf_mac = utils_sriov.get_vf_mac(sriov_test_obj.pf_name)
```
Comment (cliping) [inline libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py:70]: The same as above.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    iface_dict = sriov_test_obj.parse_iface_dict()
+    device_type = "hostdev" if dev_type == "hostdev_device" else "interface"
+    iface_dict = sriov_test_obj.parse_iface_dict()
```
Comment (cliping) [inline libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py:73]: line 71 and line 73, get iface_dict twice.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    test_dict = sriov_test_obj.parse_iommu_test_params()
+    # test_dict.update({"dev_type": 'interface', 'set_vf_mac': set_vf_mac})
+
+    br_dict = test_dict.get('br_dict', {'source': {'bridge': 'br0'}})
+    br_name = br_dict['source'].get('bridge')
+    # if not br_dict.get('mac_address'):
+    #     test_dict['br_dict']['mac_address'] = utils_net.generate_mac_address_simple()
```
Comment (cliping) [inline libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py:80]: Maybe we can remove line 79 ~ 80.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py (hunk 1 above)
Comment (Yingshun) [reply libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py]: Removed.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py (hunk 2 above)
Comment (Yingshun) [reply libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py]: Removed.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    set_vf_mac = "yes" == params.get("set_vf_mac", "no")
+    vm_name = params.get("main_vm", "avocado-vt-vm1")
+    vm = env.get_vm(vm_name)
+    sriov_test_obj = sriov_base.SRIOVTest(vm, test, params)
+    # default_vf_mac = utils_sriov.get_vf_mac(sriov_test_obj.pf_name)
+    iface_dict = sriov_test_obj.parse_iface_dict()
+    device_type = "hostdev" if dev_type == "hostdev_device" else "interface"
+    iface_dict = sriov_test_obj.parse_iface_dict()
+    test_dict = sriov_test_obj.parse_iommu_test_params()
+    # test_dict.update({"dev_type": 'interface', 'set_vf_mac': set_vf_mac})
```
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py]: Yes, done.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py (hunk 3 above)
Comment (Yingshun) [reply libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py]: Thanks! updated.

File: libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py (hunk 4 above)
Comment (Yingshun) [reply libvirt/tests/src/sriov/failover/sriov_failover_at_dt_hostdev.py]: yes, removed.

File: libvirt/tests/src/sriov/failover/sriov_failover_lifecycle.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                                               tcpdump_status_error=True)
+        check_points.check_vm_iface_num(vm_session, expr_iface_no)
+
+    dev_type = params.get("dev_type", "")
+    set_vf_mac = "yes" == params.get("set_vf_mac", "no")
+    vm_name = params.get("main_vm", "avocado-vt-vm1")
+    vm = env.get_vm(vm_name)
+    sriov_test_obj = sriov_base.SRIOVTest(vm, test, params)
+    iface_dict = sriov_test_obj.parse_iface_dict()
+    iface_dict = sriov_test_obj.parse_iface_dict()
```
[the remark this answers is not in the document]
Comment (Yingshun) [reply libvirt/tests/src/sriov/failover/sriov_failover_lifecycle.py]: updated.

**Review by cliping — APPROVED** [review]
LGTM

**Review by chloerh — APPROVED** [review]

**Comment by chloerh** [comment]
Waiting for dependency

**Pull request merged**